library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/ResourceLoaderEvents.java

```diff
@@ -0,0 +1,93 @@
+
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.resource.ServerResourceManager;
+import net.minecraft.server.MinecraftServer;
+
+import org.quiltmc.qsl.base.api.event.Event;
+import org.quiltmc.qsl.resource.loader.api.reloader.IdentifiableResourceReloader;
+
+/**
+ * Events related to the resource loader.
+ */
+public final class ResourceLoaderEvents {
+	private ResourceLoaderEvents() {
+		throw new UnsupportedOperationException("ResourceLoaderEvents only contains static definitions.");
+	}
+
+	/**
+	 * An event indicating the start of the reloading of data packs on a Minecraft server.
+	 * <p>
+	 * This event should not be used to load resources, use {@link ResourceLoader#registerReloader(IdentifiableResourceReloader)} instead.
+	 */
+	public static final Event<StartDataPackReload> START_DATA_PACK_RELOAD = Event.create(StartDataPackReload.class,
+			callbacks -> (server, resourceManager) -> {
+				for (var callback : callbacks) {
+					callback.onStartDataPackReload(server, resourceManager);
+				}
+			});
+
+	/**
+	 * An event indicating the end of the reloading of data packs on a Minecraft server.
+	 * <p>
+	 * This event should not be used to load resources, use {@link ResourceLoader#registerReloader(IdentifiableResourceReloader)} instead.
+	 */
+	public static final Event<EndDataPackReload> END_DATA_PACK_RELOAD = Event.create(EndDataPackReload.class,
+			callbacks -> (server, resourceManager, error) -> {
+				for (var callback : callbacks) {
+					callback.onEndDataPackReload(server, resourceManager, error);
+				}
+			});
+
+	/**
+	 * Functional interface to be implemented on callbacks for {@link #START_DATA_PACK_RELOAD}.
+	 *
+	 * @see #START_DATA_PACK_RELOAD
+	 */
+	@FunctionalInterface
+	public interface StartDataPackReload {
+		/**
+		 * Called before data packs on a Minecraft server have been reloaded.
+		 *
+		 * @param server             the server, may be {@code null} for the first reload
+		 * @param oldResourceManager the old resource manager, to be replaced, may be {@code null} for the first reload
+		 */
+		void onStartDataPackReload(@Nullable MinecraftServer server, @Nullable ServerResourceManager oldResourceManager);
+	}
+
+	/**
+	 * Functional interface to be implemented on callbacks for {@link #END_DATA_PACK_RELOAD}.
+	 *
+	 * @see #END_DATA_PACK_RELOAD
+	 */
+	@FunctionalInterface
+	public interface EndDataPackReload {
+		/**
+		 * Called after data packs on a Minecraft server have been reloaded.
+		 * <p>
+		 * If the reload was not successful, the old data packs will be kept.
+		 *
+		 * @param server          the server, may be {@code null} for the first reload
+		 * @param resourceManager the resource manager, may be {@code null} if the data pack reload failed
+		 * @param error           present if the data pack reload failed, otherwise {@code null}
+		 */
+		void onEndDataPackReload(@Nullable MinecraftServer server, ServerResourceManager resourceManager, @Nullable Throwable error);
```

Comment:
One of the questions I was asked when I designed an event like this in fabric api is, is the throwable important to users of the event?

Reply:
imo yes, for example on client it could redirect to a temporary screen describing the error, what is to blame, etc.

And I prefer to be able to know that something fails if the end event trigger rather than being blind on that data.